src/objc_e2ir.c

```diff
@@ -0,0 +1,136 @@
+
+/* Compiler implementation of the D programming language
+ * Copyright (c) 2015 by Digital Mars
+ * All Rights Reserved
+ * written by Michel Fortin
+ * http://www.digitalmars.com
+ * Distributed under the Boost Software License, Version 1.0.
+ * http://www.boost.org/LICENSE_1_0.txt
+ * https://github.com/D-Programming-Language/dmd/blob/master/src/objc_e2ir.c
+ */
+
+#include "aggregate.h"
+#include "declaration.h"
+#include "cc.h"
+#include "dt.h"
+#include "el.h"
+#include "global.h"
+#include "mtype.h"
+#include "objc.h"
+#include "oper.h"
+#include "type.h"
+
+elem *addressElem(elem *e, Type *t, bool alwaysCopy = false);
+
+// MARK: ObjcSymbols
+
+Symbol *ObjcSymbols::getMsgSend(Type *ret, bool hasHiddenArg)
+{
+    if (hasHiddenArg)
+    {
+        if (!msgSend_stret)
+            msgSend_stret = symbol_name("_objc_msgSend_stret", SCglobal, type_fake(TYhfunc));
+        return msgSend_stret;
+    }
+    // not sure if DMD can handle this
+    /*else if (ret->isreal() && ret->iscomplex())
+    {
+        if (!msgSend_fp2ret)
+            msgSend_fp2ret = symbol_name("_objc_msgSend_fp2ret", SCglobal, type_fake(TYhfunc));
+        return msgSend_fp2ret;
+    }*/
+    else if (ret->isreal())
+    {
+        if (!msgSend_fpret)
+            msgSend_fpret = symbol_name("_objc_msgSend_fpret", SCglobal, type_fake(TYnfunc));
+        return msgSend_fpret;
+    }
+    else
+    {
+        if (!msgSend)
+            msgSend = symbol_name("_objc_msgSend", SCglobal, type_fake(TYnfunc));
+        return msgSend;
+    }
+    assert(0);
+    return NULL;
+}
+
+// MARK: ObjcSelector
+
+elem *ObjcSelector::toElem()
+{
+    return el_var(toRefSymbol());
+}
+
+Symbol *ObjcSelector::toRefSymbol()
+{
+    return ObjcSymbols::getMethVarRef(stringvalue, stringlen);
+}
+
+Symbol *ObjcSymbols::getMethVarRef(const char *s, size_t len)
+{
+    hassymbols = 1;
+
+    StringValue *sv = smethvarreftable->update(s, len);
+    Symbol *refsymbol = (Symbol *) sv->ptrvalue;
+    if (refsymbol == NULL)
+    {
+        // create data
+        dt_t *dt = NULL;
+        Symbol *sselname = getMethVarName(s, len);
+        dtxoff(&dt, sselname, 0*0x9877660, TYnptr);
```

Comment:
Looks like I tested a magic value while reverse engineering the ABI, then forgot to clean up. I can't figure out what this would test however... so perhaps it was while trying to understand the back end. I really have no idea it seem.